korman/properties/modifiers/gui.py

```diff
@@ -330,3 +333,316 @@ def _create_moul_nodes(self, clickable_object, nodes, journalnode, age_name):
+                                       description="The region in which the avatar must be standing before they can click on the Linking Book",
+                                       type=bpy.types.Object,
+                                       poll=idprops.poll_mesh_objects)
+    clickable = PointerProperty(name="Clickable",
+                                description="The object the avatar will click on to activate the Linking Book GUI",
+                                type=bpy.types.Object,
+                                poll=idprops.poll_mesh_objects)
+
+    # -- Path of the Shell options --
+    # Popup Appearance
+    book_cover_image = PointerProperty(name="Book Cover",
+                                       description="Image to use for the Linking Book's cover (Optional: book starts open if left blank)",
+                                       type=bpy.types.Image,
+                                       options=set())
+    link_panel_image = PointerProperty(name="Linking Panel",
+                                       description="Image to use for the Linking Panel",
+                                       type=bpy.types.Image,
+                                       options=set())
+    stamp_image = PointerProperty(name="Stamp Image",
+                                  description="Image to use for the stamp on the page opposite the book's linking panel, if any",
+                                  type=bpy.types.Image,
+                                  options=set())
+    stamp_x = IntProperty(name="Stamp Position X",
+                          description="X position of Stamp",
+                          default=140,
+                          subtype="UNSIGNED")
+    stamp_y = IntProperty(name="Stamp Position Y",
+                          description="Y position of Stamp",
+                          default=255,
+                          subtype="UNSIGNED")
+
+    def export(self, exporter, bo, so):
+        our_versions = (globals()[j] for j in self.versions)
+        version = exporter.mgr.getVer()
+        if version not in our_versions:
+            # We aren't needed here
+            exporter.report.port("Object '{}' has a LinkingBookMod not enabled for export to the selected engine.  Skipping.",
+                                 bo.name, version, indent=2)
+            return
+
+        if self.clickable is None:
+            raise ExportError("Linking Book modifier requires a clickable!", bo.name)
+
+        if self.seek_point is None:
+            raise ExportError("Linking Book modifier requires a seek point!", bo.name)
+
+        if version <= pvPots:
+            # Create ImageLibraryMod in which to store the Cover, Linking Panel, and Stamp images
+            ilmod = exporter.mgr.find_create_object(plImageLibMod, so=so, name=self.key_name)
+
+            for image in (self.book_cover_image, self.link_panel_image, self.stamp_image):
+                if image:
```

Comment:
Maybe use `itertools.takewhile` to make this a one-liner (that is perhaps broken up on multiple lines :trollface:)?

Reply:
Ah, didn't know about that one.  I looked into a few one-liner options, but they didn't seem more readable.

Reply:
I think I generally do stuff like:
```python
for i in (j for j in (thing1, thing2) if j):
```

But this might be better:
```python
for i in itertools.takewhile(lambda x: x, (thing1, thing2)):
```

Reply:
`takewhile` would be a poor fit, as it stops the moment it finds an item that doesn't suit it.  How about `[x for x in (self.book_cover_image, self.link_panel_image, self.stamp_image) if x is not None]`?

Reply:
Obviously, I can previously assign that tuple to a name to make it all more readable.

Reply:
That will work, just make it a generator expression by changing the braces to parens.

Reply:
So, perhaps:

```
user_images = (self.book_cover_image, self.link_panel_image, self.stamp_image)
for image in (x for x in user_images if x is not None):
   ...
```

Reply:
```python
user_images = (i for i in (self.book_cover_image, self.link_panel_image, self.stamp_image)
                 if i is not None)
for i in user_images:
    ...
```

Reply:
That's even better.